go/zeusmaster/zeusmaster_test.go

```diff
@@ -0,0 +1,265 @@
+
+  b = sock.send($response, 0)
+  puts "Wrote #{b} bytes to #{sock_path}"
+end
+`,
+}
+
+func writeTestFiles(dir string) error {
+	for name, contents := range testFiles {
+		if err := ioutil.WriteFile(path.Join(dir, name), []byte(contents), 0644); err != nil {
+			return fmt.Errorf("error writing %s: %v", name, err)
+		}
+	}
+
+	gempath := os.Getenv("ZEUS_TEST_GEMPATH")
+	if gempath == "" {
+		var err error
+		gempath, err = filepath.Abs("rubygem")
+		if err != nil {
+			return fmt.Errorf("error finding gempath: %v", err)
+		}
+	}
+
+	if err := os.Symlink(filepath.Join(gempath, "lib"), filepath.Join(dir, "lib")); err != nil {
+		return fmt.Errorf("error linking zeus gem: %v", err)
+	}
+
+	return nil
+}
+
+func enableTracing() {
+	slog.SetTraceLogger(slog.NewTraceLogger(os.Stderr))
+}
+
+func TestZeusBoots(t *testing.T) {
+	if os.Getenv("ZEUS_LISTENER_BINARY") == "" {
+		t.Fatal("Missing ZEUS_LISTENER_BINARY env var")
+	}
+
+	dir, err := ioutil.TempDir("", "zeus_test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	if err := writeTestFiles(dir); err != nil {
+		t.Fatal(err)
+	}
+
+	config.ConfigFile = filepath.Join(dir, "zeus.json")
+	unixsocket.SetZeusSockName(filepath.Join(dir, ".zeus.sock"))
+
+	connections := map[string]*net.UnixConn{
+		"cmd":  nil,
+		"data": nil,
+		"code": nil,
+	}
+
+	for name := range connections {
+		sockName := filepath.Join(dir, fmt.Sprintf("%s.sock", name))
+
+		c, err := net.ListenUnixgram("unixgram", &net.UnixAddr{
+			Name: sockName, Net: "unixgram",
+		})
+		if err != nil {
+			t.Fatalf("Error opening %q socket: %v", sockName, err)
+		}
+		defer c.Close()
+
+		connections[name] = c
+	}
+
+	me, err := os.FindProcess(os.Getpid())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := os.Chdir(dir); err != nil {
+		t.Fatal(err)
+	}
+
+	// TODO: Find a way to redirect stdout so we can look for crashed
+	// processes.
+	enableTracing()
+	zexit := make(chan int)
+	go func() {
+		zexit <- zeusmaster.Run()
+	}()
+
+	expects := map[string]string{
+		// TODO: Use the zeusclient to spawn a command to test
+		// that path.
+		// "cmd":  "pong",
+		"data": "YAML the Camel is a Mammal with Enamel",
+		"code": "Hello, world!",
+	}
+
+	for name, want := range expects {
+		if err := readAndCompare(connections[name], want); err != nil {
+			t.Fatalf("%s: %v", name, err)
+		}
+	}
+
+	// TODO: It appears the filewatcher takes some time to initialize
+	// so we need to wait for it to propagate before changing things.
+	time.Sleep(500 * time.Millisecond)
+
+	for _, f := range []string{"data.yaml", "code.rb"} {
+		from := filepath.Join(dir, fmt.Sprintf("other-%s", f))
+		to := filepath.Join(dir, f)
+		if err := os.Rename(from, to); err != nil {
+			t.Fatalf("Error renaming %s: %v", f, err)
+		}
+	}
+
+	expects = map[string]string{
+		"data": "Hi",
+		"code": "there!",
+	}
+
+	for name, want := range expects {
+		if err := readAndCompare(connections[name], want); err != nil {
+			t.Fatalf("%s: %v", name, err)
```

Comment:
Seems there's a test failure happening around here in 2.1.7: https://travis-ci.org/burke/zeus/jobs/130500536 - seems unlikely the test it taking too long at 4s, but I guess some race condition is still happening.